### PR TITLE
[FEATURE][needs-docs][processing] add parameter representing raster band

### PR DESCRIPTION
**Pull request by alexbruy**
## Description
Add parameter to simplify raster band selection. Work in progress.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/core/processing/qgsprocessingparameters.h [hunk 1]
```diff
@@ -1849,6 +1849,58 @@ class CORE_EXPORT QgsProcessingParameterFolderDestination : public QgsProcessing
...
+ */
+class CORE_EXPORT QgsProcessingParameterBand : public QgsProcessingParameterDefinition
+{
+  public:
+
+    /**
+     * Constructor for QgsProcessingParameterBand.
+     */
+    QgsProcessingParameterBand( const QString &name, const QString &description = QString(), const QVariant &defaultValue = QVariant(),
+                                const QString &parentLayerParameterName = QString(),
```
Comment (nyalldawson) [inline src/core/processing/qgsprocessingparameters.h]: Hm - my fault, but this just made me realise that there's an inconsistency here. We should either rename this argument to `parentLayerParameter` or rename the parentLayerParameter() getter to parentLayerParameterName(). Which do you prefer?

File: src/core/processing/qgsprocessingparameters.h [hunk 2]
```diff
@@ -1849,6 +1849,58 @@ class CORE_EXPORT QgsProcessingParameterFolderDestination : public QgsProcessing
 
 };
 
+/**
+ * \class QgsProcessingParameterBand
+ * \ingroup core
+ * A raster band parameter for Processing algorithms.
+  * \since QGIS 3.0
+ */
+class CORE_EXPORT QgsProcessingParameterBand : public QgsProcessingParameterDefinition
```
Comment (nyalldawson) [inline src/core/processing/qgsprocessingparameters.h]: Don't forget to update the sip casting at the top of this header too.

**Review by nyalldawson — COMMENTED** [review]
Very nice! Can you pull across the unit tests for the field parameter and adapt too? 

Is your opinion that you'd use `self.parameterAsInt(...)` to resolve the input value here?

File: src/core/processing/qgsprocessingparameters.h (hunk 1 above)
Comment (alexbruy) [inline src/core/processing/qgsprocessingparameters.h]: IMHO better to rename getter to `parentLayerParameterName()` as it returns name of the parameter, not layer parameter itself. Do you want me to take care about it?

**Comment by alexbruy** [comment]
> Is your opinion that you'd use self.parameterAsInt(...) to resolve the input value here?

Yes, I was thinking about using `self.parameterAsInt(...)` to resolve value, as raster band combo returns int.

File: src/core/processing/qgsprocessingparameters.h (hunk 1 above)
Comment (nyalldawson) [inline src/core/processing/qgsprocessingparameters.h]: Yes please ! :)

**Comment by nyalldawson** [comment]
> Yes, I was thinking about using self.parameterAsInt(...) to resolve value, as raster band combo returns int.

:+1:

File: src/core/processing/qgsprocessingparameters.h (hunk 1 above)
Comment (alexbruy) [inline src/core/processing/qgsprocessingparameters.h]: Done

**Comment by alexbruy** [comment]
Also added tests for new parameter

**Review by nyalldawson — APPROVED** [review]
Looks good to me - is this ready to merge in your opinion?

**Comment by alexbruy** [comment]
Thanks! I think it is ready, all algorithms should be ported, tests added.

**Comment by nyalldawson** [comment]
Merge away and I'll rebase my open PR to match!

**Pull request merged**